File: acmicpc/1707/1707.cc
#include <cstdio>
#include <queue>
#include <vector>

using namespace std;

const int ARR_SIZE = 20001;
int k;
int v, e;

vector<int> t[ARR_SIZE];
short check[ARR_SIZE];

bool bfs(int s) {
    queue<int> q;
    q.push(s);
    check[s] = 1;

    while(!q.empty()) {
        int v = q.front(); q.pop();
        for (int i : t[v]) {
            if (check[i] == 0) {
                q.push(i);
                check[i] = -check[v];
            } else if (check[i] == check[v]) {
                return false;
            }
        }
    }

    return true;
}

void solve() {
    for (int i = 1; i <= v; i++) {
        if (check[i] == 0) {
            if (!bfs(i)) {
                printf("NO\n");
                return;
            }
        }
    }
    printf("YES\n");
}

int main() {
    scanf(" %d", &k);
    while (k--) {
        scanf(" %d %d", &v, &e);
        for (int i = 0; i < e; i++) {
            int v1, v2;
            scanf(" %d %d", &v1, &v2);
            t[v1].push_back(v2);
            t[v2].push_back(v1);
        }

        solve();

        for (int i = 1; i <= v; i++) {
            check[i] = 0;
            t[i].clear();
        }
    }
}

File: acmicpc/1707/1707.py
import queue

def bfs(graph, check, start):
    q = queue.Queue()
    q.put(start)
    check[start] = 1
    
    while not q.empty():
        v = q.get()
        for i in graph[v]:
            if check[i] is 0:
                q.put(i)
                check[i] -= check[v]
            elif check[i] is check [v]:
                return False
    return True

def solve(v, e):
    graph = [ [] for _ in range(v) ]
    check = [ 0 for _ in range(v) ]
    
    for _ in range(e):
        v1, v2 = [ int(i) for i in input().split() ]
        graph[v1 - 1].append(v2 - 1)
        graph[v2 - 1].append(v1 - 1)

    for i in range(v):
        if check[i] is 0:
            result = bfs(graph, check, i) 
        if result is False:
            print("NO")
            return
    print("YES")

k = int(input())

for i in range(k):
    v, e = [ int(i) for i in input().split() ]
    solve(v, e)
